### PR TITLE
Rtos porting

### PR DESCRIPTION
**Pull request by manthanand**
# Quality Assurance Checklist 
To make reviews more efficient, please make sure the software feature meets the following standards and check everything off that meets the quality check. Once everything has been checked, the assigned reviewers will begin the review process. _Edit this description to check off the list._

There are exceptions with all guidelines. As long as your decisions are justified, then you are good! Contact the reviewers or the leads about any exceptions. 

## Requirements 
- [x] Followed Coding Style Guide 
- [x] Code Build checks pass 
- [ ] No merge conflicts 
- [x] Software feature has test associated with it. 
- [x] Software feature has documentation for it updated.
- [x] Test provides useful information and uses relevant data to accurately represent BPS 
    - NOTE: If test file already exists, use that one 
- [ ] 2 Members of the team have already reviewed my PR 
    - NOTE: Approver will not look at PR if 2 other members haven't reviewed first 
- [x] Did you Have Fun?

## Things to Consider 
- [ ] Even if the above are checked, is this the best way of writing my code? 
    - It's possible to write code that works, but are there ways to make code more efficient?

### CODE REVIEW

**Comment by manthanand** [comment]
When reviewing this PR, try looking for bugs that were already in the code as well (for example deadlocks, etc.). When reviewing massive PRs like this one, I heavily rely on the diff feature of GitHub to first see what functionality was changed from before, whether it was necessary, whether the proper error checking has been done, etc.

File: Tasks/Inc/RTOS_BPS.h [hunk 1]
```diff
@@ -2,55 +2,121 @@
...
  */
-void RTOS_BPS_TimeDelay(void);
+void RTOS_BPS_SemCreate(BPS_OS_SEM* sem, char* name, uint32_t count);
 
+/**
+ * @brief Creates a Tick-based Time Delay.
+ * @param dly Defines how many ticks to delay for.
+ * @return none 
+ */
+void RTOS_BPS_DelayTick(BPS_OS_TICK dly);
```
Comment (manthanand) [inline Tasks/Inc/RTOS_BPS.h:120]: Reorder the functions in this file to be more structured (sem functions next to each other, etc.)

File: Tasks/Src/Task_Init.c [hunk 2]
```diff
@@ -125,38 +95,26 @@ void Task_Init(void *p_arg) {
 		assertOSError(err);
 		*/
```
Comment (manthanand) [inline Tasks/Src/Task_Init.c:96]: Should change the function above to meet the same parameters as the others.

File: Tasks/Src/RTOS_BPS.c [hunk 3]
```diff
@@ -4,21 +4,138 @@
 
 #include "RTOS_BPS.h"
 
-void RTOS_BPS_SemPend(void){ 
+/**
```
Comment (manthanand) [inline Tasks/Src/RTOS_BPS.c:7]: All the BPS RTOS functions should have the Fault Flag variable used (skip the function if the Fault Flag variable is set).

File: Tasks/Src/Task_Init.c [hunk 4]
```diff
@@ -15,55 +16,36 @@ void Task_Init(void *p_arg) {
...
-				10,						// Time quanta (time slice) 10 ticks
-				(void *)0,				// Extension pointer (not needed)
-				OS_OPT_TASK_STK_CHK | OS_OPT_TASK_SAVE_FP,	// Options
-				&err);					// return err code
-		assertOSError(err);
+				TASK_CRITICAL_STATE_STACK_SIZE
+		);					// return err code
         //3
-        // OSTaskCreate(&PetWDog_TCB,				// TCB
+        // RTOS_BPS_TaskCreate(&PetWDog_TCB,				// TCB
```
Comment (connorl309) [inline Tasks/Src/Task_Init.c:48]: Why is all of this commented out??

File: Tasks/Src/Task_Init.c [hunk 5]
```diff
@@ -78,38 +60,26 @@ void Task_Init(void *p_arg) {
...
-				(void *)0,				// Extension pointer (not needed)
-				OS_OPT_TASK_STK_CHK | OS_OPT_TASK_SAVE_FP,	// Options
-				&err);					// return err code
-		assertOSError(err);
+				TASK_AMPERES_MONITOR_STACK_SIZE
+				);					// return err code
 	/*
         //6
-        OSTaskCreate(&BatteryBalance_TCB,				// TCB
+        RTOS_BPS_TaskCreate(&BatteryBalance_TCB,				// TCB
```
Comment (connorl309) [inline Tasks/Src/Task_Init.c:82]: Same here

File: Tasks/Src/Task_Amperes.c [hunk 6]
```diff
@@ -68,16 +68,15 @@ void Task_AmperesMonitor(void *p_arg) {
...
 
         WDog_BitMap |= WD_AMPERES;
 
         OSMutexPost(&WDog_Mutex, OS_OPT_POST_NONE, &err);
         assertOSError(err);
         
         //delay of 10ms
-        OSTimeDly(1, OS_OPT_TIME_DLY, &err);
+        RTOS_BPS_DelayMs(10);
         assertOSError(err);
```
Comment (manthanand) [inline Tasks/Src/Task_Amperes.c:80]: Need to remove the assertOSError calls in all these files.

File: Tests/Test_CAN_Task.c [hunk 7]
```diff
@@ -94,7 +94,7 @@ void Task_Spam(void *p_arg){
         }
 
         // delay for 50ms (half the time volttemp delays for because other threads will also take CPU during the race)
-        OSTimeDly(5, OS_OPT_TIME_DLY, &err);
+        RTOS_BPS_DelayTick(5);
         assertOSError(err);
```
Comment (connorl309) [inline Tests/Test_CAN_Task.c:98]: does assertOSError need to be here?

File: Tests/Test_VoltTempMonitorTask.c [hunk 8]
```diff
@@ -157,14 +120,13 @@ void Task2(void *p_arg){
 	assertOSError(err);
 
     while(1){
-        OSMutexPend(&WDog_Mutex, 0, OS_OPT_PEND_BLOCKING, NULL, &err);
-        assertOSError(err);
+        RTOS_BPS_MutexPend(&WDog_Mutex, OS_OPT_PEND_BLOCKING);
         WDog_BitMap |= WD_AMPERES;
         WDog_BitMap |= WD_BALANCING;
         OSMutexPost(&WDog_Mutex, OS_OPT_POST_NONE, &err);
```
Comment (connorl309) [inline Tests/Test_VoltTempMonitorTask.c:126]: this didn't get changed to RTOS_BPS_MutexPost

Comment (connorl309) [reply Tests/Test_VoltTempMonitorTask.c:126]: also these assertOSErrors can be removed due to our handling that in the wrappers

File: Tests/Test_WD_Task.c [hunk 9]
```diff
@@ -23,9 +23,9 @@ void Task2(void *p_arg){
    
     while(1) {
         BSP_Light_Toggle(EXTRA);
-        OSTimeDly(25, OS_OPT_TIME_DLY, &err);
+        RTOS_BPS_DelayTick(25);
         // Comment out the following lines to test watchdog timeout
-        OSMutexPend(&WDog_Mutex, 0, OS_OPT_PEND_BLOCKING, NULL, &err);
+        RTOS_BPS_MutexPend(&WDog_Mutex, OS_OPT_PEND_BLOCKING);
         WDog_BitMap = 7;
         OSMutexPost(&WDog_Mutex, OS_OPT_POST_NONE, &err);
```
Comment (connorl309) [inline Tests/Test_WD_Task.c:30]: change

**Review by connorl309 — CHANGES_REQUESTED** [review]

File: Apps/Src/Amps.c [hunk 10]
```diff
@@ -54,27 +55,21 @@ static int32_t latestMeasureMilliAmps;
...
  * Stores and updates the new measurements received
  */
 void Amps_UpdateMeasurements(void) {
 	OS_ERR err;
-	CPU_TS ticks;
-	OSMutexPend(&AmperesData_Mutex, 0, OS_OPT_PEND_BLOCKING, &ticks, &err);
-	assertOSError(err);
+	RTOS_BPS_MutexPend(&AmperesData_Mutex, OS_OPT_PEND_BLOCKING);
 	latestMeasureMilliAmps = LTC2315_GetCurrent();
 	OSMutexPost(&AmperesData_Mutex, OS_OPT_POST_NONE, &err);
```
Comment (ray4477) [inline Apps/Src/Amps.c:74]: OSMutexPost should be replaced with wrapper version

File: Tasks/Src/Task_Init.c (hunk 4 above)
Comment (Champers5000) [reply Tasks/Src/Task_Init.c:48]: Not sure, it was originally like that when I started working on RTOS BPS task create, so I left it like that

File: Tasks/Src/Task_Init.c (hunk 5 above)
Comment (Champers5000) [reply Tasks/Src/Task_Init.c:82]: Same thing here lol

File: Tasks/Src/Task_Init.c (hunk 2 above)
Comment (Champers5000) [reply Tasks/Src/Task_Init.c:96]: It's commented out tho

Comment (manthanand) [reply Tasks/Src/Task_Init.c:96]: But it will be used later

File: Tasks/Src/Task_Init.c [hunk 11]
```diff
@@ -15,55 +16,36 @@ void Task_Init(void *p_arg) {
...
-				WATERMARK_STACK_LIMIT,	// Watermark limit for debugging
-				TASK_FAULT_STATE_STACK_SIZE,		// Stack size
-				0,						// Queue size (not needed)
-				10,						// Time quanta (time slice) 10 ticks
-				(void *)0,				// Extension pointer (not needed)
-				OS_OPT_TASK_STK_CHK | OS_OPT_TASK_SAVE_FP,	// Options
-				&err);					// return err code
-		assertOSError(err);
+				TASK_FAULT_STATE_STACK_SIZE		// Stack size
+		);					// return err code
```
Comment (manthanand) [inline Tasks/Src/Task_Init.c:37]: remove unnecessary comments

File: Tasks/Src/Task_Init.c [hunk 12]
```diff
@@ -15,55 +16,36 @@ void Task_Init(void *p_arg) {
...
-				WATERMARK_STACK_LIMIT,	// Watermark limit for debugging
-				TASK_CRITICAL_STATE_STACK_SIZE,		// Stack size
-				0,						// Queue size (not needed)
-				10,						// Time quanta (time slice) 10 ticks
-				(void *)0,				// Extension pointer (not needed)
-				OS_OPT_TASK_STK_CHK | OS_OPT_TASK_SAVE_FP,	// Options
-				&err);					// return err code
-		assertOSError(err);
+				TASK_CRITICAL_STATE_STACK_SIZE
+		);					// return err code
```
Comment (manthanand) [inline Tasks/Src/Task_Init.c:46]: follow coding styles guide (have parentheses after final parameter).

File: Tasks/Src/RTOS_BPS.c [hunk 13]
```diff
@@ -4,21 +4,138 @@
...
+void RTOS_BPS_TaskCreate(
+    BPS_OS_TCB      *p_tcb,
+    char            *p_name,
+    void            *p_task,
+    void            *p_arg,
+    uint8_t          prio,
+    BPS_CPU_STK     *p_stk_base,
+    uint64_t         stk_size
+    )
+    {
```
Comment (manthanand) [inline Tasks/Src/RTOS_BPS.c:55]: The tabs are inconsistent in this function.

**Review by connorl309 — DISMISSED** [review]
i think it all looks good - i tried checking every file but it is possible i might have missed something..

**Pull request closed without merging**